nixos/hardware.display: init module

nazarewk opened the pull request:
fixes #279739

## Description of changes

While preparing a bug report for root cause of #279739 I came up with a possible way to fix it and confirmed it is indeed working.

https://github.com/NixOS/nixpkgs/blob/580d12f7dc941cad010f3295eda1ed0cb513611c/pkgs/build-support/kernel/modules-closure.sh seems to ignore any files that are not kernel modules

As a part of the discussion feedback I have created a NixOS module out of the whole thing for handling `edid` files and additional kernel-level display configuration.

## Things done

Using it for a few weeks in my configurations https://github.com/nazarewk-iac/nix-configs

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).
---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nazarewk committed:
makeModulesClosure: include /lib/firmware/edid

modules-closure.sh seems to consider everything under
 /lib/firmware to be a kernel module,
 this change adds a special handling of `edid` directory,
 which does not contain kernel modules

fixes #279739

soupglasses reviewed (approved):
Works correctly with nice logging. Only nitpick i have is to document this specific handling of `lib/firmware/edid` somewhere. As it might not be clear to an end user that edids are required to be put into the `edid` folder to function correctly.

nazarewk commented:
> Works correctly with nice logging. Only nitpick i have is to document this specific handling of `lib/firmware/edid` somewhere. As it might not be clear to an end user that edids are required to be put into the `edid` folder to function correctly.

Yeah, I also had trouble with discovering where to put edids and how to link them together on the system, but I have no idea where to document it.

soupglasses commented:
I am not sure either. But i could see a subsection like `Using Custom EDID files` under https://nixos.org/manual/nixos/stable/#sec-kernel-config could be possible. As its very close to where `boot.initrd.kernelModules` is taught.

K900 commented:
At this point, why not a NixOS option that does all the right plumbing?

nazarewk committed:
nixos/hardware.display: init

alyssais commented on nixos/doc/manual/configuration/linux-kernel.chapter.md:
```diff
@@ -118,3 +118,74 @@ the latest kernel version that ZFS is compatible with may decrease over time.
+It is helpful in cases when either monitor-side controller or graphic driver .
+
+To make custom EDID binaries discoverable you should first create a derivation storing them at
+`/lib/firmware/edid/` and secondly add that derivation to `hardware.firmware` NixOS option.
```
```suggestion
`$out/lib/firmware/edid/` and secondly add that derivation to `hardware.firmware` NixOS option.
```

alyssais commented on nixos/doc/manual/configuration/linux-kernel.chapter.md:
````diff
@@ -118,3 +118,74 @@ the latest kernel version that ZFS is compatible with may decrease over time.
+```nix
+{
+  hardware.firmware = [
+   (pkgs.runCommandNoCC "edid-custom" {} ''
````
```suggestion
   (pkgs.runCommand "edid-custom" {} ''
```

NoCC is the default.

alyssais reviewed (commented):
module-closure.sh stuff looks fine to me

nazarewk commented:
Ok, reworked the whole thing into a NixOS module and confirmed it is working through my system configs.

MinerSebas commented on nixos/doc/manual/configuration/linux-kernel.chapter.md:
```diff
@@ -136,3 +136,85 @@ the latest kernel version that ZFS is compatible with may decrease over time.
+## Using custom EDID files {#sec-linux-config-edid-files}
+
+This section describes how to use a custom-tailored EDID file for your monitor.
+It is helpful in cases when either monitor-side controller or graphic driver .
```
when either what?
The sentence is incomplete.

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2405.section.md:
```diff
@@ -323,3 +323,5 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
 - QtMultimedia has changed its default backend to `QT_MEDIA_BACKEND=ffmpeg` (previously `gstreamer` on Linux or `darwin` on MacOS).
   The previous native backends remain available but are now minimally maintained. Refer to [upstream documentation](https://doc.qt.io/qt-6/qtmultimedia-index.html#ffmpeg-as-the-default-backend) for further details about each platform.
 
+- New module `hardware.display` streamlines working with custom display configurations allowing workarounds for misbehaving monitors
```
Please see the comment under the next headline

nazarewk replied:
You mean I should not put it at the end, but somewhere in the middle of the list?

SuperSandro2000 commented on nixos/modules/services/hardware/display.nix:
```diff
@@ -0,0 +1,141 @@
+  cfg = config.hardware.display;
+in
+{
+  ###### interface
```
Those are legacy comments and we shouldn't continue to cargo cult them.

SuperSandro2000 commented on nixos/modules/services/hardware/display.nix:
```diff
@@ -0,0 +1,141 @@
+          (pkgs.runCommand "edid-custom" {} '''
+            mkdir -p "$out/lib/firmware/edid"
+            base64 -d > "$out/lib/firmware/edid/custom1.bin" <<'EOF'
+            <insert your base64 encoded EDID file here `base64 < /sys/class/drm/card0-.../edid`>
```
We should provide an example how to download files like https://github.com/linuxhw/EDID/blob/master/Digital/ASUS/AUS3553/1F6B755EFEF5

nazarewk replied:
I've submitted a whole package doing that: https://github.com/NixOS/nixpkgs/pull/195815

nazarewk replied:
Should I pull it in maybe?

nazarewk replied:
I actually went ahead and pulled it in

SuperSandro2000 commented on nixos/modules/services/hardware/display.nix:
```diff
@@ -0,0 +1,141 @@
+        Attribute set of XFree86 Modelines automatically converted
+        and exposed as `edid/<name>.bin` files in initrd.
+        See for more information:
+        - `hardware.display.edid.packages`
```
Since this option is on the same level as this one, I don't think we need to mention it.

SuperSandro2000 commented on nixos/modules/services/hardware/display.nix:
```diff
@@ -0,0 +1,141 @@
+      '';
+      example = lib.literalExpression ''
+        {
+          "PG278Q_60" = "    241.50   2560 2608 2640 2720   1440 1443 1448 1481   -hsync +vsync";
```
Those lines get rendered on search.nixos.org and I think for that they are bit long. Maybe removing some spaces would make them better to render.

nazarewk replied:
how about this?
```
"PG278Q_60" = builtins.concatStringsSep " " [
  "241.50"
  "2560 2608 2640 2720"
  "1440 1443 1448 1481"
  "-hsync +vsync"
];
```

nazarewk replied:
but to be honest this defeats the whole purpose of being visually similar to Modelines

l0b0 replied:
> Those lines get rendered on search.nixos.org and I think for that they are bit long.

@SuperSandro2000 I'd rather have properly formatted content which overflows than improperly formatted content.

@nazarewk That said, is there a good reason to have empty space at the beginning of each modeline? We could align them further to the left.

As an aside, would tabs be appropriate to separate the pseudo-columns?

SuperSandro2000 commented on nixos/modules/services/hardware/display.nix:
```diff
@@ -0,0 +1,141 @@
+    };
+  };
+
+  ###### implementation
+
+  # services.udev implements hardware.firmware option,
+  # there is no point running EDIDs without it
+  config = lib.mkIf config.services.udev.enable {
```
Maybe it is cleaner to enabled udev below and depend on cfg.edid.packages here?

nazarewk replied:
what do you think about the new version?

nazarewk committed:
linuxhw-edid-fetcher: init at unstable-2023-05-08

l0b0 commented on nixos/modules/services/hardware/display.nix:
```diff
@@ -0,0 +1,165 @@
+            description = lib.mdDoc ''
+              An `video` kernel parameter (framebuffer mode) configuration for the specific output:
+
+                 <xres>x<yres>[M][R][-<bpp>][@<refresh>][i][m][eDd]
```
[Code blocks should be indented by 4 spaces](https://daringfireball.net/projects/markdown/syntax#precode) relative to the surrounding text.

l0b0 commented on nixos/modules/services/hardware/display.nix:
```diff
@@ -0,0 +1,165 @@
+      type = with lib.types; attrsOf (listOf str);
+      default = { };
+      description = lib.mdDoc ''
+        Exposes EDID files from https://github.com/linuxhw/EDID
```
As an end user, my first question when looking at this would be when to use it. Basically, how can I tell whether my system already has the relevant EDID file *somewhere*? I'm not certain whether this is the right place to put this documentation verbatim, but a link to a good reference would also work.

nazarewk replied:
I'm pretty sure the link contains enough background information, but I can add some practical information to the docs.

l0b0 commented on nixos/modules/services/hardware/display.nix:
```diff
@@ -0,0 +1,165 @@
+              - https://wiki.archlinux.org/title/Kernel_mode_setting#Forcing_modes
+            '';
+            example = lib.literalExpression ''
+              "e"
```
Another example with a full `<xres>x<yres>[M][R][-<bpp>][@<refresh>][i][m][eDd]` format would be helpful.

nazarewk replied:
I never used it outside "e" and don't actually know what else is it useful for. I'm pretty sure EDID always covered the use cases.

l0b0 commented on pkgs/by-name/li/linuxhw-edid-fetcher/linuxhw-edid-fetch.sh:
```diff
@@ -0,0 +1,94 @@
+
+main() {
+  if [[ $# == 0 ]] ; then
+    usage
```
Nit: since this is an error condition, the usage message should probably go to stderr:

```suggestion
    usage >&2
```

l0b0 commented on pkgs/by-name/li/linuxhw-edid-fetcher/linuxhw-edid-fetch.sh:
```diff
@@ -0,0 +1,94 @@
+    exit 1
+  fi
+
+  repo="${repo:-"$PWD"}"
+
+  if [[ -d "$1" ]]; then
+    repo="$1"
+    shift
+  fi
```
I'd prefer if this assumed the synopsis on line 9, rather than allowing a default:
```suggestion
  repo="$1"
  shift
```

The checks on line 65 will take care of verifying that `$1` points to a directory.

nazarewk replied:
I actually intend it to default to `${src}` for a package, but also want it to make sense on it's own (outside Nix).

l0b0 replied:
Having two separate synopses, one which has to be detected by a hack (checking if the first argument is a directory) isn't ideal though. What if you call `linuxhw-edid-fetcher PG278Q 2014 >edid.bin`, and there happens to be a `PG278Q` directory? It would do the wrong thing. Better to do something explicit, like

```bash
if [[ $# -eq 3 ]]; then
    do the directory thing
elif [[ $# -eq 2 ]]; then
    do the non-directory thing
else
    usage >&2
    exit 2
fi
```

nazarewk replied:
counting arguments definitely does not make sense in here as the search can have arbitrary number of arguments depending on the specific device. I actually expected `PG278Q` to be enough for mine, but it turns out there are over 10 flavors right now (used to be ~4 when I originally written the script)

nazarewk replied:
reworked the thing a bit not to use the argument at all and giving example on overriding the directory using `repo=/path/to/EDID`

l0b0 reviewed (commented):
Love this addition, just have a few minor suggestions.

l0b0 commented on pkgs/by-name/li/linuxhw-edid-fetcher/package.nix:
```diff
@@ -0,0 +1,66 @@
+    mkdir -p "$out/bin"
+    ln -s "$fetch" "$out/bin/"
+    ${lib.optionalString (displays != { }) ''
+      install -Dm 444 *.bin -t "$out/lib/firmware/edid"
```
Most Linux CLIs (and especially GNU ones) expect the positional arguments *last.* Also, it's hard to know what the `D`, `m`, and `t` flags mean unless you're very familiar with this command, so I'd recommend using the long options (where available). Taken together:

```suggestion
      install -D --mode=444 --target-directory="$out/lib/firmware/edid" *.bin
```

(unfortunately, there's no long option for `-D`.)

nazarewk replied:
To be honest I don't even know where the `install` command comes from, just copy pasting it from existing derivations.

nazarewk commented:
thanks to https://github.com/NixOS/nixpkgs/blob/8da441b7a71f401af0b57f0480c912e219cddb08/nixos/modules/services/networking/netbird.nix#L38-L42 I seem to have found a proper way to document a new module

meutraa commented:
Can we get this merged? 

I've been using it since the inception without problems.

MrMonotone commented:
Seems to fail when using the example for linuxhw. Line 31: prints: - : invalid option

MrMonotone commented on pkgs/by-name/li/linuxhw-edid-fetcher/linuxhw-edid-fetch.sh:
```diff
@@ -0,0 +1,96 @@
+        ;;
+    *)
+        log "Matched entries:"
+        log "- %s" "${lines[@]}"
```
This should be escaped \-

MrMonotone commented on pkgs/by-name/li/linuxhw-edid-fetcher/linuxhw-edid-fetch.sh:
```diff
@@ -0,0 +1,96 @@
+        ;;
+    1)
+        log "Matched entries:"
+        log "- %s" "${lines[@]}"
```
This should be escaped \-

MrMonotone commented:
> Can we get this merged?
> 
> I've been using it since the inception without problems.

I have had several issues. I am able to get to a desktop by plugging and unplugging my screen sometimes? Only can get to desktop with nomodeset usually. Perhaps the other option is working well? How do you use it?

config
```
  # Fix display
  hardware.display.edid.linuxhw = { PG278Q_2014 = [ "PG278Q" "2014" ]; };
  hardware.display.edid.modelines."PG278Q_60" = "   241.50   2560 2608 2640 2720   1440 1443 1448 1481   -hsync +vsync";
  hardware.display.edid.modelines."PG278Q_120" = "  497.75   2560 2608 2640 2720   1440 1443 1448 1525   +hsync -vsync";
  hardware.display.outputs."DP-1".edid = "PG278Q_60.bin";
  hardware.display.outputs."DP-1".mode = "e";
  ```


logs repeatedly pumping out:
```
Mar 03 05:31:57 nixos kernel: [drm] Initialized nouveau 1.4.0 20120801 for 0000:26:00.0 on minor 0
Mar 03 05:31:57 nixos kernel: nouveau 0000:26:00.0: Direct firmware load for edid/PG278Q_60.bin failed with error -2
Mar 03 05:31:57 nixos kernel: nouveau 0000:26:00.0: [drm] *ERROR* [CONNECTOR:43:DP-1] Requesting EDID firmware "edid/PG278Q_60.bin" failed (err=-2)
Mar 03 05:31:57 nixos kernel: nouveau 0000:26:00.0: [drm] Cannot find any crtc or sizes
```

meutraa commented:
I only use the edid firmware file option, not the modeline to edid generator.

MrMonotone commented:
> I only use the edid firmware file option, not the modeline to edid generator.

Could you show me how you edid firmware option? The example provided does not work.

meutraa commented:
> > I only use the edid firmware file option, not the modeline to edid generator.
> 
> Could you show me how you edid firmware option? The example provided does not work.

```
          hardware.display = {
            outputs."DP-1".edid = "edid.bin";
            edid.packages = [
              (pkgs.runCommand "edid" { } ''
                mkdir -p $out/lib/firmware/edid
                base64 -d > "$out/lib/firmware/edid/edid.bin" <<'EOF'
AP///////wBB0P8JAAAAAAAAAQMIAAB46u6Vo1RMmSYPUFQhCACBwIEAgYCVAKnAswABAQEB79wAGPE4QEAwIDoAwA4xAAAaAAAAEAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAkUCAzBARJA/BAMjCQcHgwEAAOMFwwFtGgAAAgcw8AAEYAJLAuYGBQFgSgDlAYuEkAFWXgCgoKApUDAgNQCAaCEAABpvwgCgoKBVUDAgNQCAaCEAABo06YAYcTgzQGwgNQDgDhEAABpHawAY8TgfQDAgOgDADjEAABoAAAAAAAAAz3ASeQAAAwFkXvYCgP8TPwF/gB8AnwUuACAABwDCgQEA/xMXAS+AHwCfBVMAAgAJAG+7AAD/ExcBL4AfAJ8FKAACAAkAStUBAP8OFwEvgB8ANwSGAAIACQCvlAEA/wk/AX+AHwCfBToAAgAEAAAAAAAAAAAAAAAAAAAAAAAAACqQ
                EOF
              '')
            ];
          };
```

Also, if anyone finds this in the future, that is a edid with working 240hz for a samsung odyssey g9 oled using amdgpu.

Which makes the correct kernel parameter.

```
initrd=\efi\nixos\gwsmjgzw4ddcmqsawk4g0v91f487dj9b-initrd-linux-6.7.6-initrd.efi init=/nix/store/w7a8isyl62gpfwa358yc7jynk7zbzwly-nixos-system-samara-24.05.20240226.d5be97c/init amdgpu.seamless=1 splash loglevel=4 drm.edid_firmware=DP-1:edid/edid.bin
```

dmesg doesn't show any relevant information when it loads successfully, I think. But I used to get that it failed to load (because it was compressed) before this PR.

MrMonotone commented:
> > > I only use the edid firmware file option, not the modeline to edid generator.
> > 
> > 
> > Could you show me how you edid firmware option? The example provided does not work.
> 
> ```
>           hardware.display = {
>             outputs."DP-1".edid = "edid.bin";
>             edid.packages = [
>               (pkgs.runCommand "edid" { } ''
>                 mkdir -p $out/lib/firmware/edid
>                 base64 -d > "$out/lib/firmware/edid/edid.bin" <<'EOF'
> AP///////wBB0P8JAAAAAAAAAQMIAAB46u6Vo1RMmSYPUFQhCACBwIEAgYCVAKnAswABAQEB79wAGPE4QEAwIDoAwA4xAAAaAAAAEAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAkUCAzBARJA/BAMjCQcHgwEAAOMFwwFtGgAAAgcw8AAEYAJLAuYGBQFgSgDlAYuEkAFWXgCgoKApUDAgNQCAaCEAABpvwgCgoKBVUDAgNQCAaCEAABo06YAYcTgzQGwgNQDgDhEAABpHawAY8TgfQDAgOgDADjEAABoAAAAAAAAAz3ASeQAAAwFkXvYCgP8TPwF/gB8AnwUuACAABwDCgQEA/xMXAS+AHwCfBVMAAgAJAG+7AAD/ExcBL4AfAJ8FKAACAAkAStUBAP8OFwEvgB8ANwSGAAIACQCvlAEA/wk/AX+AHwCfBToAAgAEAAAAAAAAAAAAAAAAAAAAAAAAACqQ
>                 EOF
>               '')
>             ];
>           };
> ```
> 
> Also, if anyone finds this in the future, that is a edid with working 240hz for a samsung odyssey g9 oled using amdgpu.
> 
> Which makes the correct kernel parameter.
> 
> ```
> initrd=\efi\nixos\gwsmjgzw4ddcmqsawk4g0v91f487dj9b-initrd-linux-6.7.6-initrd.efi init=/nix/store/w7a8isyl62gpfwa358yc7jynk7zbzwly-nixos-system-samara-24.05.20240226.d5be97c/init amdgpu.seamless=1 splash loglevel=4 drm.edid_firmware=DP-1:edid/edid.bin
> ```
> 
> dmesg doesn't show any relevant information when it loads successfully, I think. But I used to get that it failed to load (because it was compressed) before this PR.

I don’t have an initrd parameter could that be the issue? Is the firmware saved in that?

MrMonotone commented:
I believe I have narrowed down the issue. I think it has to deal with having a LUKS partition. I boot my system, load nouveau, black screen, unplug monitor and plug it back in, enter password to decrypt LUKS, Screen comes back on boot. I have to unplug my monitor before entering the password or it does not work. Do you guys have LUKS partitions?

nazarewk commented:
> I believe I have narrowed down the issue. I think it has to deal with having a LUKS partition. I boot my system, load nouveau, black screen, unplug monitor and plug it back in, enter password to decrypt LUKS, Screen comes back on boot. I have to unplug my monitor before entering the password or it does not work. Do you guys have LUKS partitions?

I am running from ZFS on top of LUKS and it works just fine and should not interfere with setting up displays. The configuration is part of initrd/unencrypted boot partition and kernel boot parameters, so it is activated long before decrypting disks

1. Are you sure you have 2014 batch of Asus PG278Q monitor?
2. The one issue I have is that sometimes after sleep/suspend I need to unplug the monitor from power (not just turn off with a button, luckily I have smart power socket, so I do that with an app instead of physically unplugging) and plug back in.
3. You seem to have nvidia card, the specific configuration I works for AMD GPU. For AMD forcing `e` skips the failing code path in the driver as per [the linked issue](https://gitlab.freedesktop.org/drm/amd/-/issues/615#note_1987392), but you might have to locate and tweak something else for nvidia (or even noveau specifically).

MrMonotone commented:
> I am running from ZFS on top of LUKS and it works just fine and should not interfere with setting up displays. The configuration is part of initrd/unencrypted boot partition and kernel boot parameters, so it is activated long before decrypting disks
The only reason why I think this is because the display loses output right when it requests my password but it shows loading the nouveau in the initrd etc.
> 1. Are you sure you have 2014 batch of Asus PG278Q monitor?
I believe so
```
  Model name............... ROG PG278Q
  Windows description...... Generic PnP Monitor
  Manufacturer............. Asus
  Manufacture date......... 2014, ISO week 39
  Filter driver............ None
```
I extracted this from my windows install. I have the rest of the information if I might need a custom edid. I think I would need help to figure that out.
> 3. You seem to have nvidia card, the specific configuration I have works for AMD GPU. For AMD forcing `e` skips the failing code path in the driver as per [the linked issue](https://gitlab.freedesktop.org/drm/amd/-/issues/615#note_1987392), but you might have to locate and tweak something else for nvidia (or even noveau specifically).
From my understanding the function referenced, drm_get_edid, is from the drm source. I believe that both amd and nouveau use this code because they both are part of the drm source tree. 

https://gitlab.freedesktop.org/drm/nouveau/-/blob/nouveau-next/drivers/gpu/drm/nouveau/nouveau_connector.c#L588

Is where it is called followed by the error:

https://gitlab.freedesktop.org/drm/nouveau/-/blob/nouveau-next/drivers/gpu/drm/nouveau/nouveau_connector.c#L592

Since you have spent more time on this would you mind advising me on what I could try? Also, how would I go about reporting this appropriately because it seems like no longer just an amd tree issue.

nazarewk commented:
> Since you have spent more time on this would you mind advising me on what I could try? Also, how would I go about reporting this appropriately because it seems like no longer just an amd tree issue.

I'm not sure myself, I guess you could try to search for existing or create a new noveau issue and link it with the amd issue

MrMonotone commented:
> > Since you have spent more time on this would you mind advising me on what I could try? Also, how would I go about reporting this appropriately because it seems like no longer just an amd tree issue.
> 
> I'm not sure myself, I guess you could try to search for existing or create a new noveau issue and link it with the amd issue

Okay I will give a shot. Not forcing the monitor to enable gave me a complete black screen. At least I know now that the solution partly works. I am just in a worse situation! Hopefully since the issue is spreading to both sets of hardware you might get a permanent fix!

meutraa commented:
This has stopped working for me since the recent commits. I've been trying to get it working all day.

nazarewk commented:
> This has stopped working for me since the recent commits. I've been trying to get it working all day.
> 
> [drm] _ERROR_ [CONNECTOR:93:DP-1] Requesting EDID firmware "edid/edid.bin" failed (err=-22)

Can you paste the whole configuration responsible for EDID? Including registration of edid files and their usage on kernel commandline?

meutraa commented:
When I am trying support-initrd-edid only:

```
{
  inputs = { nixpkgs.url = "github:nazarewk/nixpkgs/support-initrd-edid"; };
    
  outputs = inputs@{ self, nixpkgs, ... }: {
    nixosConfigurations = {
      samara = nixpkgs.lib.nixosSystem {
        system = "x86_64-linux";
        modules = [
          ({ pkgs, ... }: {
            services.udev.enable = true;
            boot.kernelParams = [ "drm.edid_firmware=DP-1:edid/g9.bin" ];
            hardware.firmware = [
              (pkgs.runCommand "edid" { } ''
                mkdir -p $out/lib/firmware/edid
                base64 -d > "$out/lib/firmware/edid/g9.bin" <<'EOF'
                AP///////wBB0P8JAAAAAAAAAQMIAAB46u6Vo1RMmSYPUFQhCACBwIEAgYCVAKnAswABAQEB79wAGPE4QEAwIDoAwA4xAAAaAAAAEAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAkUCAzBARJA/BAMjCQcHgwEAAOMFwwFtGgAAAgcw8AAEYAJLAuYGBQFgSgDlAYuEkAFWXgCgoKApUDAgNQCAaCEAABpvwgCgoKBVUDAgNQCAaCEAABo06YAYcTgzQGwgNQDgDhEAABpHawAY8TgfQDAgOgDADjEAABoAAAAAAAAAz3ASeQAAAwFkXvYCgP8TPwF/gB8AnwUuACAABwDCgQEA/xMXAS+AHwCfBVMAAgAJAG+7AAD/ExcBL4AfAJ8FKAACAAkAStUBAP8OFwEvgB8ANwSGAAIACQCvlAEA/wk/AX+AHwCfBToAAgAEAAAAAAAAAAAAAAAAAAAAAAAAACqQ
                EOF
              '')
            ];
          })
        ];
      };
    };
  };
}
```

And my original configuration:

```
{
  inputs = { 
    nixpkgs.url = "github:nazarewk/nixpkgs/hardware-firmware-edid-fix";
  };

  outputs = inputs@{ self, nixpkgs, ... }: {
    nixosConfigurations = {
      samara = nixpkgs.lib.nixosSystem {
        system = "x86_64-linux";
        modules = [
          ({ pkgs, ... }: {
            hardware.display = {
              outputs."DP-1".edid = "edid.bin";
              edid.packages = [
                (pkgs.runCommand "edid" { } ''
                  mkdir -p $out/lib/firmware/edid
                  base64 -d > "$out/lib/firmware/edid/edid.bin" <<'EOF'
                  AP///////wBB0P8JAAAAAAAAAQMIAAB46u6Vo1RMmSYPUFQhCACBwIEAgYCVAKnAswABAQEB79wAGPE4QEAwIDoAwA4xAAAaAAAAEAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAkUCAzBARJA/BAMjCQcHgwEAAOMFwwFtGgAAAgcw8AAEYAJLAuYGBQFgSgDlAYuEkAFWXgCgoKApUDAgNQCAaCEAABpvwgCgoKBVUDAgNQCAaCEAABo06YAYcTgzQGwgNQDgDhEAABpHawAY8TgfQDAgOgDADjEAABoAAAAAAAAAz3ASeQAAAwFkXvYCgP8TPwF/gB8AnwUuACAABwDCgQEA/xMXAS+AHwCfBVMAAgAJAG+7AAD/ExcBL4AfAJ8FKAACAAkAStUBAP8OFwEvgB8ANwSGAAIACQCvlAEA/wk/AX+AHwCfBToAAgAEAAAAAAAAAAAAAAAAAAAAAAAAACqQ
                  EOF
                '')
              ];
            };
          })
        ];
      };
    };
  };
}
```

Both of these used to work. I used the top one before you turned the PR into a hardware module, and the bottom one after.

I've reverted to my original solution before this PR which still does work, it is just not as nice:

```
{
  systemd.services.edid = let
    edid =
      "AP///////wBB0P8JAAAAAAAAAQMIAAB46u6Vo1RMmSYPUFQhCACBwIEAgYCVAKnAswABAQEB79wAGPE4QEAwIDoAwA4xAAAaAAAAEAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAkUCAzBARJA/BAMjCQcHgwEAAOMFwwFtGgAAAgcw8AAEYAJLAuYGBQFgSgDlAYuEkAFWXgCgoKApUDAgNQCAaCEAABpvwgCgoKBVUDAgNQCAaCEAABo06YAYcTgzQGwgNQDgDhEAABpHawAY8TgfQDAgOgDADjEAABoAAAAAAAAAz3ASeQAAAwFkXvYCgP8TPwF/gB8AnwUuACAABwDCgQEA/xMXAS+AHwCfBVMAAgAJAG+7AAD/ExcBL4AfAJ8FKAACAAkAStUBAP8OFwEvgB8ANwSGAAIACQCvlAEA/wk/AX+AHwCfBToAAgAEAAAAAAAAAAAAAAAAAAAAAAAAACqQ";
  in {
    enable = true;   
    script = ''    
      sleep 3
      echo ${edid} | base64 -d > /sys/kernel/debug/dri/1/DP-1/edid_override
      echo 1 >/sys/kernel/debug/dri/1/DP-1/trigger_hotplug
    '';
    wantedBy = [ "graphical.target" ];
  };   
}
```

nazarewk commented:
I [updated my configs](https://github.com/nazarewk-iac/nix-configs/commit/7490b3f02d96ae0fbe41fff48642c5eca3b1f9c4) and confirmed it is properly putting files into intird:
```
[I] kdn@krul ~/d/g/n/nix-configs (main)> sudo lsinitrd $(sudo find /boot/loader/entries -maxdepth 1 -name "*.conf" -print | sort | sudo xargs cat | grep initrd | tail -n 1 | sed 's#initrd /efi#/boot/EFI#Ig') | grep 'edid'
Please touch the device.
drwxr-xr-x   0 root     root            0 Jan  1  1970 ./nix/store/pylw9gazw2s53wyyi8zc83n625b4gfrb-linux-6.7.11-modules-shrunk/lib/firmware/edid/
-r--r--r--   0 root     root          128 Jan  1  1970 ./nix/store/pylw9gazw2s53wyyi8zc83n625b4gfrb-linux-6.7.11-modules-shrunk/lib/firmware/edid/PG278Q_120.bin
-r--r--r--   0 root     root          128 Jan  1  1970 ./nix/store/pylw9gazw2s53wyyi8zc83n625b4gfrb-linux-6.7.11-modules-shrunk/lib/firmware/edid/PG278Q_60.bin
-r--r--r--   0 root     root          128 Jan  1  1970 ./nix/store/pylw9gazw2s53wyyi8zc83n625b4gfrb-linux-6.7.11-modules-shrunk/lib/firmware/edid/U2711_60.bin
```
then rebooted and it keeps working.

I'm using systemd-boot, systemd initrd and this piece of config:
```
hardware.display.edid.modelines."PG278Q_120" = "   497.75   2560 2608 2640 2720   1440 1443 1448 1525   +hsync -vsync";
hardware.display.outputs."DP-1".edid = "PG278Q_120.bin";
hardware.display.outputs."DP-1".mode = "e";
```

@meutraa can you try checking your intird with `sudo lsinitrd $(sudo find /boot/loader/entries -maxdepth 1 -name "*.conf" -print | sort | sudo xargs cat | grep initrd | tail -n 1 | sed 's#initrd /efi#/boot/EFI#Ig') | grep 'edid'` ? it basically tries to locate latest `systemd-boot` entry, read it's `intird` path and list files using dracut's `lsinitrd`.

nazarewk commented:
seems like https://discourse.nixos.org/t/copying-custom-edid/31593/26

nazarewk commented:
@meutraa thanks for the idea, I have added `hardware.display.edid.applyAtRuntime` and `hardware.display.edid.applyWithKernelParameters` to work around `drm.edid_firmware=` failures such as those happening right now.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/copying-custom-edid/31593/27

meutraa commented on nixos/modules/services/hardware/display.nix:
```diff
@@ -167,13 +189,60 @@ in
+            export PATH="${lib.makeBinPath (with pkgs; [ findutils coreutils ])}:$PATH"
+
+            edid_paths=(
+              /run/current-system/firmware/edid
```
```
$ find /run/current-system/firmware/edid/ -maxdepth 1 -name edid.bin
/run/current-system/firmware/edid/edid.bin

$ find /run/current-system/firmware/edid -maxdepth 1 -name edid.bin

$ ls /run/current-system/firmware/edid
edid.bin
```

The paths in edid_paths have to end with a / for maxdepth 1 to work. I tested this with normal files and it is not true, but for some reason it is true with /run/current-system/firmware/edid. Some weirdness because of it being a symlink maybe.

As a result: 

Applying '' to '/sys/kernel/debug/dri/0000:03:00.0/DP-1' output
cat: '': No such file or directory

nazarewk replied:
added `/` at the end of each entry

nazarewk commented:
FYI: after today's nixos-unstable update (bump from last week) my Asus PG278Q became completely invisible

nazarewk commented:
> FYI: after today's nixos-unstable update (bump from last week) my Asus PG278Q became completely invisible to the system

Actually I had to run this manually after system booted (those 2 hotplugs might be redundant, but I'm not sure):
```
echo on >/sys/kernel/debug/dri/1/DP-1/force
echo 1 >/sys/kernel/debug/dri/1/DP-1/trigger_hotplug
cat /run/current-system/firmware/edid/PG278Q_120.bin >/sys/kernel/debug/dri/1/DP-1/edid_override
echo 1 >/sys/kernel/debug/dri/1/DP-1/trigger_hotplug
```

so the kernel stopped handling `video=DP-1:e`

SuperSandro2000 commented on pkgs/by-name/li/linuxhw-edid-fetcher/package.nix:
```diff
@@ -0,0 +1,67 @@
+}:
+
+# Usage:
+#   let
+#     edids = (linuxhw-edid-fetcher.override {
+#       displays = {
+#         PG278Q_2014 = [ "PG278Q" "2014" ];
+#       };
+#     });
+#   in
+#     "${edids}/lib/firmware/edid/PG278Q_2014.bin";
```
```suggestion
#   let
#     edids = linuxhw-edid-fetcher.override {
#       displays.PG278Q_2014 = [ "PG278Q" "2014" ];
#     };
#   in
#     "${edids}/lib/firmware/edid/PG278Q_2014.bin";
```
we can minimize that example a bit further, to make it easier to understand

nazarewk replied:
actually I added (unnecessary) `2560x1440` to make it obvious `PG278Q_2014.bin` is derived from the object key, not a search path.

SuperSandro2000 commented on pkgs/by-name/li/linuxhw-edid-fetcher/package.nix:
```diff
@@ -0,0 +1,67 @@
+    ''}
+  '';
+
+  passthru.updateScript = nix-update-script { extraArgs = [ "--version=branch=master" ]; };
```
```suggestion
  passthru.updateScript = nix-update-script { extraArgs = [ "--version=branch=master" ]; };

```

SuperSandro2000 commented on pkgs/by-name/li/linuxhw-edid-fetcher/linuxhw-edid-fetch.sh:
```diff
@@ -0,0 +1,96 @@
+    test -d "$1" && test -f "$1/AnalogDisplay.md" && test -f "$1/DigitalDisplay.md"
+}
+
+main() {
```
We could just drop the main function

nazarewk replied:
I prefer to have an explicit entry point in form of `main()` because:
- hard to follow scripts interweaving running code with function definitions
- it's a lot harder to locate where the code starts executing

SuperSandro2000 commented on pkgs/by-name/li/linuxhw-edid-fetcher/linuxhw-edid-fetch.sh:
```diff
@@ -0,0 +1,96 @@
+        repo="${TMPDIR:-/tmp}/edid"
+        log "Not running inside 'https://github.com/linuxhw/EDID', downloading content to ${repo}"
+        if ! check_repo "$repo"; then
+            curl -L https://github.com/linuxhw/EDID/tarball/master | tar -zx -C "${repo}" --strip-components=1
```
```suggestion
            curl -L https://github.com/linuxhw/EDID/tarball/HEAD | tar -zx -C "${repo}" --strip-components=1
```
just in case they ever change the default branch

nazarewk replied:
nice, wasn't aware of this

SuperSandro2000 commented on pkgs/by-name/li/linuxhw-edid-fetcher/linuxhw-edid-fetch.sh:
```diff
@@ -0,0 +1,96 @@
+        exit 1
+    fi
+
+    : "${repo:="$PWD"}"
```
Shouldn't we rather use a tmp file for this?

nazarewk replied:
it's just for detecting whether current directory is already checked out repository, if it's not it downloads stuff to `/tmp/edid`

SuperSandro2000 commented on nixos/modules/services/hardware/display.nix:
```diff
@@ -0,0 +1,254 @@
+        };
+      }
+    )
+    (lib.mkIf (cfg.edid.packages != null) {
```
We should probably hide this behind en enable flag. Some modules, eg in nixos-hardware might fill this and having to use mkForce to clear it, is not that nice.

SuperSandro2000 reviewed (commented):
I have no clue about edit but looking pretty good to my naive eyes and we can probably merge this with those small cleanups.

nazarewk commented:
@SuperSandro2000 thanks for taking interest in it, I'm currently working on a separate script for doing the runtime configuration and would LOVE to merge after I got it to work.

Yesterday I updated nixos-unstable in my configs + swapped Radeon RX550 to Pro W6600 and the current script stopped working completely because kernel/`amdgpu` started doing additional wild stuff like turning off display after failing to load EDID even though it was force enabled. Needed to add a few additional guards against it.

nazarewk committed:
edido: init

nazarewk commented:
ok, I've created and [tested](https://github.com/nazarewk-iac/nix-configs/commit/048a47256c8275a7b2afaa88e8ed23e0855845ec) in my configs `edido` script which applies (selected) `video=` and complete `drm.edid_firmware=` kernel parameters during runtime. It loads `/proc/cmdline` automatically unless you pass it arguments directly.

meutraa commented:
I have issues with the latest version of the display-edid-apply script.

With applyWithKernelParameters false, the only log line displayed is:

```
log "loading kernel parameters from /proc/cmdline"
```

In this case the edid is not applied.

---

With applyWithKernelParameters true it ends with:

```
log "EDID is already up to date with '%s'" "${edidPath}"
```

I checked the hash of the current-firmware edid.bin and it did indeed match that of the edid_override. So CHANGED remains 0, and the trigger hotplug doesn't occur. With a manual trigger hotplug, the edid is functioning. I'm really not sure what is going on here. Perhaps something odd with the kernel where it loads it, but doesn't really.

nazarewk commented:
Yeah, the kernel has major issues loading EDID files over last few weeks, just getting worse with each update.

Good catch with having no `boot.kernelParams` unless `applyWithKernelParameters` is set.

billksun commented:
Never thought I would need this, but yes, I need this!

jys1670 commented:
Seems that errors like
`amdgpu 0000:e7:00.0: [drm] *ERROR* [CONNECTOR:111:DP-5] Requesting EDID firmware "edid/edid.bin" failed (err=-22)`
are caused by https://github.com/NixOS/nixpkgs/commit/1ed5df1e703135d2d6814a761ce67e6e8073af30 (and unrelated to kernel version bumps)
With its revert, `applyWithKernelParameters = true;` works fine on my machine.

jys1670 commented:
> Complete shot in the dark, but if you use the unmodified kernel and add `initcall_blacklist=simpledrm_platform_driver_init`, does it work?

Nope, I can't even boot with this one, sadly.

FlyingWombat commented on nixos/modules/services/hardware/display.nix:
```diff
@@ -0,0 +1,235 @@
+          pathsToLink = [ "/lib/firmware/edid" ];
+          ignoreCollisions = true;
+        }) // {
+          compressFirmware = false;
```
We shouldn't actually need `compressFirmware = false;`, since the kernel is built with compressed firmware support by default. But it looks trivial either way: trivial space savings with compression (for edid bins) vs. added complexity in nixos config.

nazarewk replied:
it is needed, I think code paths in kernel/drivers do not decompress EDID files or at least not properly. Don't remember the details, but i tried to investigate quite extensively (without much knowledge of C/C++ itself)

FlyingWombat replied:
But my archlinux partition is using a zstd compressed edid firmware file without issue. Using Archlinux vanilla kernel v6.8.9.arch1-2.
```
$ cat /proc/cmdline
BOOT_IMAGE=/boot/vmlinuz-linux root=UUID=eb3729f4-547c-42a7-86b4-6020f9a31762 rw loglevel=3 quiet nowatchdog intel_iommu=on iommu=pt zswap.enabled=0 ibt=off amdgpu.ppfeaturemask=0xffffffff drm.edid_firmware=edid/g70nc_edid.bin

$ ls /usr/lib/firmware/edid/
g70nc_edid.bin.zst
```
I compressed it manually with `zstd -19 g70nc_edid.bin`

FlyingWombat replied:
@nazarewk  With 1ed5df1e703135d2d6814a761ce67e6e8073af30 reverted, my nixos partition loads zstd compressed edid firmware ( as of nixos-unstable commit ad57eef4ef0659193044870c731987a6df5cf56b , kernel v6.9.3).

```
$ file /run/current-system/firmware/edid/g70nc_edid.bin.zst
/run/current-system/firmware/edid/g70nc_edid.bin.zst: Zstandard compressed data (v0.8+), Dictionary ID: None
```

Here is the relevant nixos config:
```nix
{ config, lib, pkgs, ... }:
let
  samsung-G70NC-edid = pkgs.runCommand "g70nc_edid.bin" { } ''
    mkdir -p $out/lib/firmware/edid
    cp ${./samsung_G70NC_edid_fix.bin} $out/lib/firmware/edid/g70nc_edid.bin
  '';
in {
  boot.kernelParams = [
    "intel_iommu=on" "iommu=pt"
    "zswap.enabled=0"  # because using zram
    "amdgpu.ppfeaturemask=0xffffffff"  # unlock amdgpu voltage and clock adjustments
    "ibt=off"  # for bcachefs
    "drm.edid_firmware=edid/g70nc_edid.bin"
  ];
  hardware.firmware = [
    samsung-G70NC-edid
  ];
}
```

FlyingWombat commented:
> Seems that errors like `amdgpu 0000:e7:00.0: [drm] *ERROR* [CONNECTOR:111:DP-5] Requesting EDID firmware "edid/edid.bin" failed (err=-22)` are caused by [1ed5df1](https://github.com/NixOS/nixpkgs/commit/1ed5df1e703135d2d6814a761ce67e6e8073af30) (and unrelated to kernel version bumps) With its revert, `applyWithKernelParameters = true;` works fine on my machine.

That's odd that commit is causing the error, since the kernel configs in that commit match what is set in my Archlinux partition (dual boot), and it loads edid_firmware (compressed) without issue. It must be failing in combination with some other options nixos is setting.

FlyingWombat commented:
I noticed Archlinux has [this patch](https://github.com/archlinux/linux/commit/c3c26154d77465013db79d1a3fad0fd2563257f0) in their kernel, to "skip simpledrm if nvidia-drm.modeset=1 is set". It could be relevant to the issues that @jys1670 mentioned were introduced with 1ed5df1e703135d2d6814a761ce67e6e8073af30 -- if the user has an Nvidia gpu.

Though, I have an AMD gpu and still get the same err=-22 when loading edid firmware.

jys1670 commented:
> Seems that errors like `amdgpu 0000:e7:00.0: [drm] *ERROR* [CONNECTOR:111:DP-5] Requesting EDID firmware "edid/edid.bin" failed (err=-22)` are caused by [1ed5df1](https://github.com/NixOS/nixpkgs/commit/1ed5df1e703135d2d6814a761ce67e6e8073af30) (and unrelated to kernel version bumps) With its revert, `applyWithKernelParameters = true;` works fine on my machine.

Can someone check if `CONFIG_FW_LOADER=y` resolves the issue (without revert of [1ed5df1](https://github.com/NixOS/nixpkgs/commit/1ed5df1e703135d2d6814a761ce67e6e8073af30))? This way EDID stuff works for me with simpledrm, but extra confirmations are appreciated.
```diff
diff --git a/pkgs/os-specific/linux/kernel/common-config.nix b/pkgs/os-specific/linux/kernel/common-config.nix
index c9bf29616062..ee05f70610e0 100644
--- a/pkgs/os-specific/linux/kernel/common-config.nix
+++ b/pkgs/os-specific/linux/kernel/common-config.nix
@@ -1126,6 +1126,8 @@ let
       # For systemd-binfmt
       BINFMT_MISC   = option yes;
 
+      FW_LOADER = yes;
+
       # Disable the firmware helper fallback, udev doesn't implement it any more
       FW_LOADER_USER_HELPER_FALLBACK = option no;
```

mrbaguettovic commented:
> Can someone check if `CONFIG_FW_LOADER=y` resolves the issue (without revert of [1ed5df1](https://github.com/NixOS/nixpkgs/commit/1ed5df1e703135d2d6814a761ce67e6e8073af30))?

Worked for me, thank you! Took 1 hour to compile but worth it.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/copying-custom-edid/31593/32

FlyingWombat commented:
@jys1670 That change worked for me too, without reverting simpledrm. Loading edid works; zstd compression works too.

jys1670 commented:
Thanks for confirmation!
@nazarewk could you incorporate that kernel config change into this PR? I've asked @K900, it's fine to include that fix in one PR, with careful merge afterward.

nazarewk commented:
> Thanks for confirmation! @nazarewk could you incorporate that kernel config change into this PR? I've asked @K900, it's fine to include that fix in one PR, with careful merge afterward.

I can simply include this patch as a commit in the PR, but I won't have access to the affected computer until ~25th June.

nazarewk committed:
kernel: fix EDID firmware loading

for context:
- https://github.com/NixOS/nixpkgs/pull/279789#discussion_r1624936659
- https://github.com/NixOS/nixpkgs/pull/279789#issuecomment-2148560802
- https://github.com/NixOS/nixpkgs/pull/279789#issuecomment-2150726766

K900 commented:
Can we drop all the userspace shenanigans now that we have the kernel working properly?

nazarewk commented:
> Can we drop all the userspace shenanigans now that we have the kernel working properly?

I'm up for it as soon as I have access to the affected computer (end of June) to test the solution end-to-end.

jys1670 commented:
Yeah, apply at runtime workaround should probably be removed.
I've tried it on KDE Plasma, it's likely overriding too late (or does not do hotplug). I see incorrect refresh rate in display settings, despite `display-edid-apply.service` finishing without errors. Correct values appear after manual `echo 1 > /sys/kernel/debug/dri/.../trigger_hotplug`

nazarewk commented:
> Yeah, apply at runtime workaround should probably be removed. I've tried it on KDE Plasma, it's likely overriding too late (or does not do hotplug). I see incorrect refresh rate in display settings, despite `display-edid-apply.service` finishing without errors. Correct values appear after manual `echo 1 > /sys/kernel/debug/dri/.../trigger_hotplug`

Yeah I've also observed it sometimes not working, seems like it needs to settle in before hotplug works depending on the display manager/something else.

nazarewk commented:
FYI: I'm back home, but my desktop drive died in [pretty common way for this model](https://old.reddit.com/r/buildapc/comments/rbgjm7/xpg_gammix_s70_blade_nvme_ssd_issue/?sort=new) so it will take me a few days more before I get it running again and be able to test the trimmed down setup.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/config-build-fails-due-to-boot-initrd-extrafiles-permission-denied/35976/4

jys1670 commented:
@nazarewk  do you need any help? I would really like to see your work merged sooner. I have been using it for, like, two months or so, and it's working as expected. The only thing that irritates me is the time-consuming kernel rebuilds 😅

As for testing stuff, you probably don't even need a machine with a malfunctioning monitor: edid fields like the week/year of manufacturing or serial number can be changed (with a final checksum correction, edid-decode prints a valid value), and a modded file can be used as a test subject. But considering how many people replied to/referenced this PR, I doubt any serious problems will arise with the overall approach. This just needs a little cleanup from ` applyAtRuntime ` hack, and it's basically ready to be merged.

nazarewk commented:
> @nazarewk do you need any help? I would really like to see your work merged sooner. I have been using it for, like, two months or so, and it's working as expected. The only thing that irritates me is the time-consuming kernel rebuilds 😅

I'll get this PR through after finishing the most immediate work task.

Thought I'd need the original desktop to finish it (I REALLY want to make sure it works end to end) and for some weird reason the screen registers properly when connected to laptop through [this HUB](https://zenwire.eu/pl/p/HUB-USB-C-10w1-Display-Port-1.4-HDMI-2.1-8K-4K-120Hz-60Hz-Full-HD-144Hz-USB-3.0-SD-Power-Delivery-100W-Macbook-M1-M2-Zenwire/288), but I just got myself a "dumb USB-C to DP" adapter which replicates the same issue on the laptop so it's available for testing.

nazarewk commented:
OK, so I have:
- dropped the runtime apply
- kept the `edido` script
- [confirmed it's working](https://github.com/nazarewk-iac/nix-configs/commit/255e6213de9686bccc6ef016438faa7c40dce30e) with USB-C <> DP adapter in my laptop

jys1670 commented:
Thanks! Updated stuff works fine on latest unstable in my use case (tiansu monitors connected via usb4 <> dp adapter)
Let's poke @K900 once again to speed things up

K900 commented on pkgs/by-name/ed/edido/package.nix:
```diff
@@ -0,0 +1,19 @@
+    "/run/current-system/firmware"
+  ]
+}:
+writeShellApplication {
```
This whole thing isn't necessary anymore, is it?

nazarewk replied:
I would like to keep it, to initially test config on a running system before rebooting, but I could simply move it to my own repo.

K900 replied:
I think it's fine if it's just for testing.

K900 commented:
It's been way too freaking long. Let's merge this and iterate later. Force-pushed to resolve the conflict, as discussed with @nazarewk on Matrix. Thanks everyone!

ciferkey commented:
Thank you all so much for the hard work on this! I had a BenQ XL2420G that worked with nvidia but on not AMD. Confirmed that its all good now with:

```
  hardware.display.edid.linuxhw."XL2420G_2014" = ["XL2420G" "2014"];
  hardware.display.outputs."DP-2".edid = "XL2420G_2014.bin";
  hardware.display.outputs."DP-2".mode = "e";
```
